### PR TITLE
glsl_shader_gen: Use epsilon for both ends of NDC range

### PR DESCRIPTION
**Pull request by GPUCode**
Aimed to fix https://github.com/citra-emu/citra/issues/7352

Part of the sky dome is rendered with depth values a little lower the -1 lower NDC range of the 3DS (The exact depth value generated is -1.00000214061). So it ends up being clipped and the clipped area showed as black.

The solution is to apply an epsilon check for this as well. A larger epsilon was chosen for this case as the depth value generated is a bit further away. I've done some hardware testing and it appears one must multiply r=-1.f by 0.1f about 5 times before a triangle with depth set to -1.f + r becomes visible again, so it seems this change matches hardware behavior.

Also refactored the clipping into a separate function so it's a little cleaner and less duplicated all over the place

- - -
This change is [Reviewable](https://reviewable.io/reviews/citra-emu/citra/7355)

### CODE REVIEW

**Comment by GPUCode** [comment]
In an ideal world this would have fixed the Rune Factory 4 lake bug as well, but the game is sneaky that the depth goes over the limit only after depth_offset is added, so we would need to either add a similar check in the fragment shader, or handle the pica transformation in the vertex shader

File: src/video_core/shader/generator/glsl_shader_gen.cpp [hunk 1]
```diff
@@ -94,13 +107,7 @@ std::string GenerateTrivialVertexShader(bool use_clip_planes, bool separable_sha
     out += GetVertexInterfaceDeclaration(true, use_clip_planes, separable_shader);
     out += VSUniformBlockDef;
 
-    // Certain games render 2D elements very close to clip plane 0 resulting in very tiny
```
Comment (FearlessTobi) [inline src/video_core/shader/generator/glsl_shader_gen.cpp:97]: we should keep this comment around somewhere

**Comment by vitor-k** [comment]
I was looking at the android support channel and saw this by coincidence; it seems to crash when using opengles, but not vulkan (adreno):
```
[  11.686283] Render.OpenGL <Error> video_core/renderer_opengl/gl_shader_util.cpp:LoadShader:68: Error compiling vertex shader:
ERROR: 0:38: '=' :  assigning non-constant to 'const float'
ERROR: 1 compilation errors.  No code generated.


[  11.686306] Render.OpenGL <Error> video_core/renderer_opengl/gl_shader_util.cpp:LoadShader:69: Shader source code:
#version 320 es

#if defined(GL_ANDROID_extension_pack_es31a)
#extension GL_ANDROID_extension_pack_es31a : enable
#endif // defined(GL_ANDROID_extension_pack_es31a)

#if defined(GL_EXT_clip_cull_distance)
#extension GL_EXT_clip_cull_distance : enable
#endif // defined(GL_EXT_clip_cull_distance)
layout(location = 0) in vec4 vert_position;
layout(location = 1) in vec4 vert_color;
layout(location = 2) in vec2 vert_texcoord0;
layout(location = 3) in vec2 vert_texcoord1;
layout(location = 4) in vec2 vert_texcoord2;
layout(location = 5) in float vert_texcoord0_w;
layout(location = 6) in vec4 vert_normquat;
layout(location = 7) in vec3 vert_view;
out vec4 primary_color;
out vec2 texcoord0;
out vec2 texcoord1;
out vec2 texcoord2;
out float texcoord0_w;
out vec4 normquat;
out vec3 view;

#ifdef VULKAN
layout (set = 0, binding = 1, std140) uniform vs_data {
#else
layout (binding = 1, std140) uniform vs_data {
#endif
    bool enable_clip1;
    vec4 clip_coef;
};

const vec2 EPSILON_Z = vec2(0.00000001f, -1.00001f);

vec4 SanitizeVertex(vec4 vtx_pos) {
    const float ndc_z = vtx_pos.z / vtx_pos.w;
    if (ndc_z > 0.f && ndc_z < EPSILON_Z[0]) {
        vtx_pos.z = 0.f;
    }
    if (ndc_z < -1.f && ndc_z > EPSILON_Z[1]) {
        vtx_pos.z = -vtx_pos.w;
    }
    return vtx_pos;
}

void main() {
    primary_color = vert_color;
    texcoord0 = vert_texcoord0;
    texcoord1 = vert_texcoord1;
    texcoord2 = vert_texcoord2;
    texcoord0_w = vert_texcoord0_w;
    normquat = vert_normquat;
    view = vert_view;
    vec4 vtx_pos = SanitizeVertex(vert_position);
    gl_Position = vec4(vtx_pos.x, vtx_pos.y, -vtx_pos.z, vtx_pos.w);

        gl_ClipDistance[0] = -vtx_pos.z; // fixed PICA clipping plane z <= 0
        if (enable_clip1) {
            gl_ClipDistance[1] = dot(clip_coef, vtx_pos);
        } else {
            gl_ClipDistance[1] = 0.0;
        }
        }
```

File: src/video_core/shader/generator/glsl_shader_gen.cpp (hunk 1 above)
Comment (GPUCode) [reply src/video_core/shader/generator/glsl_shader_gen.cpp:97]: I feel this comment is too specific to account for the other cases handled by this PR. Given that using epsilons for clipping checks isn't something that esoteric I would imagine a reader would easily figure out it's for handling these precision differences between the host and the guest

**Comment by GPUCode** [comment]
The mali error should be fixed now, though it's pretty sus that their compiler won't allow that

**Review by Steveice10 — APPROVED** [review]

**Pull request merged**